syntax = "proto2";
package oneflow;

import "oneflow/core/job/job.proto";
import "oneflow/core/job/resource.proto";
import "oneflow/core/job/file_system_conf.proto";

message IOConf {
  required FileSystemConf data_fs_conf = 1;
  required FileSystemConf snapshot_fs_conf = 2;
  optional bool save_downloaded_file_to_local_fs = 3 [default = false];
  optional uint64 persistence_buf_byte = 4 [default = 67108864]; // 64M
}

message EpollConf {
}

message IBVerbsConf {
  optional string device_name = 1 [default = ""];
  optional uint32 port_num = 2 [default = 0];
  optional uint32 sgid_index = 3 [default = 0];
  optional uint32 pkey_index = 4 [default = 0];
  optional uint32 queue_depth = 5 [default = 1024];
  optional uint32 timeout = 6 [default = 14];
  optional uint32 retry_cnt = 7 [default = 7];
  optional uint32 service_level = 8 [default = 0];
  optional uint32 traffic_class = 9 [default = 0];
  optional uint64 mem_block_mbyte = 10 [default = 8];
}

message CommNetworkConf {
  oneof comm_net_type {
    EpollConf epoll_conf = 1;
    IBVerbsConf ibverbs_conf = 2;
  }
}

message ProfilerConf {
  optional bool collect_act_event = 1 [default = false];
}

message ReuseMemPriorityStrategy {
}

message ParallelismPriorityStrategy {
}

message JobNameGroup {
  repeated string job_name = 1;
}

message CustomParallelismStrategy {
  repeated JobNameGroup nonparallel_group = 1;
}

message InterJobReuseMemStrategy {
  oneof strategy_case {
    ReuseMemPriorityStrategy reuse_mem_priority = 1;
    ParallelismPriorityStrategy parallelism_priority = 2;
    CustomParallelismStrategy custom_parallelism = 3;
  }
}

message ConfigProto {
  required Resource resource = 1;
  required IOConf io_conf = 2;
  required CommNetworkConf comm_net_conf = 3;
  optional ProfilerConf profiler_conf = 4;
  repeated string load_lib_path = 5;
}

message JobSet {
  repeated Job job = 1;
  optional InterJobReuseMemStrategy inter_job_reuse_mem_strategy = 5;
}
